### ISSUE TITLE
ssr: TC resolution in have

### ISSUE THREAD

**Issue by gares** [coq/coq#6094]
@vbgl and @bgregoir report that the current choice about TC resolution in `have` is suboptimal.
Example:
```
have <ipat> := lemma args ?TC.
```
In this case, as per doc, ?TC is not solved but generalized. Note that ?TC does not occur in the type of `lemma args ?TC` and that `?TC : SomeClass ... : Prop`, so it is doubly irrelevant. Not resolving ?TC forces the ipat to do some gimmicks, like `/(_ _)` to force the "specialization", or to pass some
extra `_` later on.

I see two possible ways to tune the behavior:
1. a syntactic flag to say "please resolve TCs"
2. the heuristics above (always look for instances of TC not occurring in the type and being in Prop)

When we added some TC support in ssr, @cohencyril and @ggonthier took part to the design, so I'm putting them in the loop.

FTR there is an easy workaround in the meanwhile (implementing 1. with a best effort syntax):
```
Notation "!! x" := (ltac:(refine x)) (at level 100, only parsing)
```
that lets one write
```
have ... := !! f args...
```
and have TC resolved.

**Comment by herbelin** [comment]
Didn't we have plans to provide what you maybe call 1. above, namely, variants of `?[n]` telling to force or prevent type class resolution? The main thing which is refraining me up to now is to find a good syntax for it, which is neither too cryptic nor too verbose.

Or why not `!!` while at it?

Or maybe `?!`, or `?[!!]` to keep a symbolic connection with the `?[n]` notation? Or maybe `?[+]` to force type class resolution and `?[-]` to prevent it?

**Comment by gares** [comment]
A dedicated syntax would work for `_` one writes, but typically TC instances are `_` automatically inserted by Coq, so I don't think this is an option.  Note that `!!` in my example is morally a flag to have, not a flag for the instance ?TC.

**Comment by CohenCyril** [comment]
Actually, I like your workaround better than a flag for have, what more would it bring?
And if people are unhappy with with the syntax, its easy for them to change it in their own development, at least it can be added to ssr for documentation purposes and I don't think it conflicts with anything in the math-comp library.

**Comment by herbelin** [comment]
> A dedicated syntax would work for _ one writes, but typically TC instances are _ automatically inserted by Coq, so I don't think this is an option. Note that !! in my example is morally a flag to have, not a flag for the instance ?TC.

Maybe could you give the exact example you had in mind then( since the one that you gave suggested that you had an explicit ?TC evar)? That would be nice from you. As for `_` inserted automatically, it is indeed difficult to imagine a general rule since I guess it would depend on the kind of instance under consideration.

**Comment by mattam82** [comment]
It could be that implicit instance arguments automatically get elaborated
to ?_ or whatever, I think that’s the solution adopted by Agda for example.
Le mar. 7 nov. 2017 à 11:22, Hugo Herbelin <notifications@github.com> a
écrit :

**Comment by herbelin** [comment]
@mattam82: Hum, I think I'm missing context to understand what you mean. What solution does Agda propose?

**Comment by mattam82** [comment]
As far as I understand, instance arguments are marked with {{ }} in Agda, and typeclass resolution is only applied to metavariables introduced by those arguments.

**Comment by herbelin** [comment]
Statically then? So, are you suggesting that statically marking argument to be resolved by type class instantiation would be enough in general, and that some use-time `?[noTC]` or `?[useTC]` is not "necessarily necessary"?

(To be honest, I don't have a lot of experience with type classes, so I'm missing examples to form a judgement of what would be optimal to do.)

**Comment by mattam82** [comment]
I think this modality should be provided to the user as well. I guess we could even generalize to a kind of « scope » or tagging of evars that could allow different resolution strategies. @CohenCyril has use-cases for that I believe, and it would make @ppedrot happy I’m sure to disentangle typeclass resolution from refinement.

**Comment by herbelin** [comment]
So, you mean that both a form of static `{{ }}` and dynamic `?[noTC]`/`?[TC]` (or maybe a syntax to be elaborated on the idea`?{{}}`, by syntactic similarity?) would be useful, right?

What would be the specification? Would the `noTC` flag take precedence over the `no_goals_or_obligations` and `no_goals` filters? And the `TC` flag overwrite, or make a conjunction` with `no_goals_or_obligations`?

**Comment by mattam82** [comment]
Yes, I mean both static and dynamic. I think with such a tagging system in place, evars for goals would be tagged by "goal" meaning notc (but maybe the default tactic should apply instead), with a variant "maybegoal" for tentative resolution (i.e., allowed to fail and leave the goal unsolved, the current semantics for obligations) and another "typeclass" for mandatory resolution for example.

**Comment by herbelin** [comment]
Thanks, I'll meditate about it.

**Comment by gares** [comment]
The example is something like
```
Lemma foo x y {p : SomeClass x y} : P x y.
..
Lemma that_uses_foo_in_have a b (q : SomeClass a b) : goal.
have := foo a b.
```
Now results in a new goal:
```
(SomeClass a b -> P a b) -> goal.
```
while the desired behavior is
```
P a b -> goal.
```
since the context item `q` is a valid instance of `SomeClass a b`.
Note that the third argument of `foo` is maximally inserted.

**Comment by herbelin** [comment]
Thanks a lot for the example. Is it a correct intuition that we want this one to be always resolved (and not generalized)?

Do you have other typical examples where it is a generalization which is expected?

**Comment by gares** [comment]
> Actually, I like your workaround better than a flag for have, what more would a flag bring?

in this simple scenario not much, but for more complex calls to have like the following one
```
have : type := !! term.
```
ssr would normally refine a cast `(term : type)` to help the pretyper, while with the !! hack things are done in different orders and I've no clue what tactic in term does for `( ltac:(refine term) : type )`.

**Comment by gares** [comment]
Moreover, since `refine term` tries to fill TC, it better have accumulated all constraints.  
Concrete example on top of my recent message:
```
have : P a b := foo _ _.
```
you don't want to resolve the TC before having understood that the two keys are `a` and `b`, so you really want to the cast and `refine` to "commute".

**Comment by mattam82** [comment]
(ltac:(refine term) : type) uses the typing constraint.

**Comment by gares** [comment]
Oh, minor fix to the workaround: use `simple refine` otherwise unresolved implicits are not accepted.
